src/server/scripts/Outland/zone_hellfire_peninsula.cpp

```diff
@@ -1039,6 +1041,212 @@ class npc_magister_aledis : public CreatureScript
+                        break;
+
+                    for (Creature* infernal : infernalrainList)
+                        if (!infernal->isMoving() && infernal->GetPositionZ() > 118.0f)
+                            infernal->AI()->SetData(DATA_ACTIVE, DATA_ACTIVE);
+
+                    break;
+                }
+                case EVENT_FEAR_CONTROLLER_CAST:
+                {
+                    std::list<Creature*> fearcontrollerList;
+                    Trinity::AllCreaturesOfEntryInRange checkerFear(me, NPC_FEAR_CONTROLLER, 200.0f);
+                    Trinity::CreatureListSearcher<Trinity::AllCreaturesOfEntryInRange> searcherFear(me, fearcontrollerList, checkerFear);
+                    Cell::VisitAllObjects(me, searcherFear, 200.0f);
+
+                    if (fearcontrollerList.empty())
+                        break;
+
+                    for (Creature* fearController : fearcontrollerList)
+                        fearController->AI()->SetData(DATA_ACTIVE, DATA_ACTIVE);
+
+                    break;
+                }
+                case EVENT_ACTIVE_FALSE:
+                    _events.ScheduleEvent(EVENT_LEONUS_TALK, 1h);
+                    _events.ScheduleEvent(EVENT_INFERNAL_RAIN_ATTACK, 1h);
+                    _events.ScheduleEvent(EVENT_FEAR_CONTROLLER_CAST, 1h);
+                    break;
+            }
+        }
+
+        if (!UpdateVictim())
+            return;
+
+        DoMeleeAttackIfReady();
+    }
+
+private:
+    EventMap _events;
+};
+
+enum InfernalRainHellfire
+{
+    EVENT_INFERNAL_RAIN_CAST   = 1,
+    EVENT_INFERNAL_RAIN_STOP   = 2,
+    NPC_WATCH_COMMANDER_LEONUS = 19392
+};
+
+struct npc_infernal_rain_hellfire : public ScriptedAI
+{
+    npc_infernal_rain_hellfire(Creature* creature) : ScriptedAI(creature) { }
+
+    void SetData(uint32 /*type*/, uint32 data) override
+    {
+        switch (data)
+        {
+            case 1:
+                _events.ScheduleEvent(EVENT_INFERNAL_RAIN_CAST, 1s, 2s);
+                _events.ScheduleEvent(EVENT_INFERNAL_RAIN_STOP, 60s);
+                break;
+            default:
+                break;
+        }
+    }
+
+    void UpdateAI(uint32 diff) override
+    {
+        _events.Update(diff);
+
+        while (uint32 eventId = _events.ExecuteEvent())
+        {
+            switch (eventId)
+            {
+                case EVENT_INFERNAL_RAIN_CAST:
+                {
+                    std::list<Creature*> infernalrainList;
+                    Trinity::AllCreaturesOfEntryInRange checker(me, NPC_INFERNAL_RAIN, 200.0f);
+                    Trinity::CreatureListSearcher<Trinity::AllCreaturesOfEntryInRange> searcher(me, infernalrainList, checker);
+                    Cell::VisitAllObjects(me, searcher, 200.0f);
+
+                    if (infernalrainList.empty())
+                        break;
+
+                    if (Creature* random = Trinity::Containers::SelectRandomContainerElement(infernalrainList))
```

Comment:
list is not empty, so Trinity::Containers::SelectRandomContainerElement(infernalrainList) will never return a null value
so you dont need check it, just:
```cpp
Creature* random = Trinity::Containers::SelectRandomContainerElement(infernalrainList);
 if (random->isMoving() && random->GetPositionZ() < 118.0f)
    me->CastCustomSpell(SPELL_INFERNAL_RAIN, SPELLVALUE_MAX_TARGETS, 1, random, true);
```

Reply:
@Keader - what you are saying is not correct. The list does not contain nullptr as a value inside it (a grid searcher will NEVER push nullptr to result list). What happens when you use `Trinity::Containers::SelectRandomContainerElement` on empty container is something else. It will return (*list.end()) which is not equal to nullptr